elastic/kibana
# Ask LLM to select relevant documents as part of recall

## Summary

This PR modifies the `recall` function from this:

- Grab user prompt and/or queries
- Pass input to Elasticsearch for semantic search using ELSER to find at most 5 hits
- Return hits

to this:

- Grab user prompt and/or queries
- Pass input to Elasticsearch for semantic search using ELSER (we could ask for many more hits here)
- Take hits and pass them to the LLM with a new system prompt asking it to score the documents based on relevancy (by forcing it to call a function)
- Sort the ELSER hits by LLM score and grab the top 5

### How to test
_It might be easier to test this once it has been merged with https://github.com/elastic/kibana/pull/171933_
_Or simply edit [this](https://github.com/elastic/kibana/blob/2abfa54accafd603f63a7e0b72b59ba6e5f65148/x-pack/plugins/observability_ai_assistant/public/functions/recall.ts#L107) line to inject your own list of documents_

- Put a large number of entries into the Knowledge base (ideally more than 20)
- Ask the Assistant questions that relate to some of those entries more than others
- Check that the initial recall HTTP request returns hits
- Check that the result of recall given to the LLM seem relevant
